source/features/useful-not-found-page.tsx

```diff
@@ -107,17 +62,133 @@ function init(): false | void {
+		const currentBranch = url.branch;
+		const {filePath} = url;
+		if (!currentBranch || !filePath) {
+			return;
+		}
+
+		const commits = await api.v3(`commits?path=${filePath}&sha=${currentBranch}&per_page=2`);
+		if (!commits[0]) {
+			return;
+		}
+
+		const commitInfo = await api.v3(`commits/${commits[0].sha as string}`);
+		const fileInfo = commitInfo.files.find((file: AnyObject) => [file.filename, file.previous_filename].includes(filePath));
+		if (!fileInfo) {
+			return;
+		}
 
-	if (['tree', 'blob'].includes(parts[2])) {
-		// Object might be 410 Gone
-		void addCommitHistoryLink(bar);
+		const commitSha = commitInfo.sha.slice(0, 8);
+		const commitDate = commitInfo.commit.committer.date;
+		const linkToCommit = commitInfo.html_url;
+
+		url.assign({
+			route: 'commits',
+		});
+		const linkToCommitHistory = url.toString();
+
+		if (fileInfo.status === 'removed') {
+			return {
+				type: 'removed',
+				commitDetails: {
+					filePath,
+					commitSha, commitDate,
+					linkToCommit, linkToCommitHistory,
+					linkToLastVersion: fileInfo.blob_url,
+				},
+			};
+		}
+
+		if (fileInfo.status === 'renamed') {
+			url.assign({
+				route: 'blob',
+				branch: commitInfo.parents[0].sha,
+				filePath,
+			});
+
+			return {
+				type: 'renamed',
+				commitDetails: {
+					filePath: fileInfo.previous_filename, newFilePath: fileInfo.filename,
+					commitSha, commitDate,
+					linkToCommit, linkToCommitHistory,
+					linkToLastVersion: url.toString(), linkToNewVersion: fileInfo.blob_url,
+				},
+			};
+		}
+	})();
+
+	if (change) {
+		if (change.type === 'renamed') {
+			bar.after(
+				<p className="container mt-4 text-center">
+					<a href={change.commitDetails.linkToLastVersion}>This file</a> was renamed to <a href={change.commitDetails.linkToNewVersion}>{change.commitDetails.newFilePath}</a> (<a href={change.commitDetails.linkToCommit}><relative-time datetime={change.commitDetails.commitDate}/></a>) - view the file&apos;s <a href={change.commitDetails.linkToCommitHistory}>commit history</a>.
+				</p>,
+			);
+			return;
+		}
+
+		if (change.type === 'removed') {
+			bar.after(
+				<p className="container mt-4 text-center">
+					<a href={change.commitDetails.linkToLastVersion}>This {getType()}</a> was removed (<a href={change.commitDetails.linkToCommit}><relative-time datetime={change.commitDetails.commitDate}/></a>) - view the file&apos;s <a href={change.commitDetails.linkToCommitHistory}>commit history</a>.
+				</p>,
+			);
+		}
 	}
+}
 
-	if (['tree', 'blob', 'edit'].includes(parts[2])) {
-		// File might not be available on the current branch
-		// GitHub already redirects /tree/ and /blob/ natively
-		void addDefaultBranchLink(bar);
+function init(): void {
+	if (
+		parseCurrentURL().length <= 1 || !select.exists('[alt*="This is not the web page you are looking for"]')
+	) {
+		return;
 	}
+
+	const bar = <h2 className="container mt-4 text-center"/>;
+	void showMissingPart(bar);
+	void showHelpfulLinks(bar);
```

Comment:
I think it is easier to `select` the `#parallax_illustration` element and keep appending using `bar.after`. Otherwise we could give an id (maybe `missing-parts-bar`) to our bar element and and add append elements below that. What do you think?

Reply:
Nah, forget `bar`, it should just be used in one place. Every time we add an element we just use: `select('main').append(x)` or something like that. I only happens twice in the whole file anyway.

Reply:
Done! I used `select('h2')!.after(...)` instead of `select('main')!.after(...)` to keep the text right below the bar.